### ISSUE TITLE
Make Android / iOS version of AdGuard not to filter Telnet Bulletin Board System (BBS). (Feature request)

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/CoreLibs#184]
## This issue needs to be re-tested when we implement CoreLibs in the Android app

@lancelot-moon commented on [Mon Nov 20 2017](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601)

@ameshkov
@vozersky

I think it's better to open a new issue for the feature request.
https://github.com/AdguardTeam/AdguardForAndroid/issues/1520

ameshkov:
> Should we disable HTTPS filtering for these apps by default?

lancelot-moon:
> I disabled AdGuard's HTTPS filtering in advance.
And then I tried to login into PTT BBS via the default port setting 443 without SSH connection.
If using JPTT, I have to wait for 2 minutes for login.
The waiting time is too long.

> If using Mo PTT, I was unable to connect to PTT BBS server.
Mo PTT showed the error,
"Oh! The server has no response, please try again later."

> Disabling HTTPS filtering is absolutely not the solution.
A risk of being unable to connect to PTT BBS.

Привет ameshkov,

Here is the information about PTT BBS.

https://github.com/ptt/pttbbs
If you have advanced questions, you can try to ask them there.
https://github.com/ptt/pttbbs/issues

> mainly used in telnet / ssh protocol, currently web-interface is used only for reading articles, no other interactions.

> SSH connection: bbs@ptt.cc
Telnet unsafe connection: telnet://ptt.cc
Website: www.ptt.cc

AdGuard is unable to filter SSH connection, right?
How does AdGuard deal with Telnet protocol?
Telnet protocol:
https://ru.wikipedia.org/wiki/Telnet

The problem is unsafe Telnet connection with AdGuard.
I tried to disable HTTPS filtering.
But it's useless.
There is still the risk of being unable to connect to PTT BBS if using the default port 443 for mobile BBS apps.
You can try Mo PTT for login into PTT BBS as a guest with the default port 443 by yourself.
Please click "我沒有帳號，路過看看".
It means:
"I don't have an account. Visit as a guest."
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/vjUAbPU.png)
</details><br/>

I don't know if not filtering all Telnet BBS is a good thing.
However, at least, can you design that AdGuard doesn't filter the 3 servers of PTT Telnet BBS by default?
If using the default port 443 in those PTT app with AdGuard, I'm usually unable to login into PTT.
I hope that I can login into PTT even if I don't modify the default port 443 in those PTT apps to 23 or 3000~3010.
PTT Telnet BBS has the 3 associated servers.
telnet://ptt.cc <- (批踢踢實業坊) (Main server)
telnet://ptt2.cc <- (批踢踢兔) 
telnet://ptt3.cc <- (批踢踢參)

The allowed ports for connecting to PTT BBS are only 23, 443 and 3000~3010.
For example:
telnet://ptt.cc:443
telnet://ptt.cc:3005

> No, that's because AG does not filter connections to ports less than 80

Does it seem that AG also does not filter connections to ports above 3000?

The general default port setting of telnet is 23.

But most of mobile BBS apps use port 443 as the default port setting, such as JPTT, Mo PTT...etc.

A few BBS apps don't use unsafe Telnet connection.
They use SSH connection as the only way of connecting to PTT BBS, such as PTT+, PiTT, isPTT, PTTopia...etc.

Спасибо!

---

@ameshkov commented on [Thu Nov 30 2017](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-348271174)

@lancelot-moon we'll solve it in v2.11, sorry for the late reply

---

@lancelot-moon commented on [Thu Nov 30 2017](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-348278235)

Taiwanese net friends thank you very much!!!
The allowed ports for connecting to PTT BBS are only 23, 443 and 3000~3010.
If not filtering Telnet protocol or by default simply not filtering telnet://ptt.cc, telnet://ptt2.cc and telnet://ptt3.cc, it's a great news!

---

@lancelot-moon commented on [Fri Dec 01 2017](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-348315906)

Hi,
Some Taiwanese people are trying new safe connection to PTT BBS.
https://github.com/iamchucky/PttChrome

An experimental safe connection to PTT BBS, compared with SSH.
It's WebSocket Protocol.
We won't need a BBS app.
Only use a browser for login into PTT BBS.
https://term.ptt.cc

I worry if AdGuard also filters WebSocket Protocol and causes the connection issue.
AdGuard is unable to filter SSH.
AdGuard should be unable to filter WebSocket since WebSocket is also a safe connection, right?
Just a share. XD

---

@ameshkov commented on [Mon Dec 04 2017](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-348823769)

No problem with WebSocket as it works on top of the regular HTTPS.

---

@Revertron commented on [Fri Mar 02 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-369959560)

Hm, I've installed Mo PTT, and if I click on "我沒有帳號，路過看看" I'm able to see many topics and posts. I don't speak Japanese, but I think it works perfectly from here (in root-proxy mode).

---

@lancelot-moon commented on [Fri Mar 02 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-369969053)

Hi,

It's Traditional Chinese, not Japanese.

I rarely use local proxy mode because my phone has no root.

Maybe you should use local VPN mode for the test.

If there is a PTT BBS account for login, users are difficult even unable to login into PTT BBS by default port 443 with AdGuard VPN when using a mobile PTT BBS app via unsafe BBS Telnet connection.
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/XqJvy7r.png)
</details><br/>

Need to modify port number to 23 or 3000~3010 in mobile PTT BBS app because AdGuard doesn't filter those port numbers.
Or simply use SSH connection.

If there is no account and then enter as a guest, I don't know why Mo PTT app's users can smoothly enter PTT BBS by default port 443 with AdGuard.

Maybe you need a PTT BBS account and then you'll understand my meaning.

Do you have an email or skype?
I maybe need to register a new account for AdGuard team's test.

---

@Revertron commented on [Fri Mar 02 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-369971411)

Yes, you can send me a letter to revertron (at) adguard.com with some credentials and I will check it.

---

@lancelot-moon commented on [Fri Mar 02 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-369996527)

@Revertron

Hi,

I have sent a PTT BBS id & pw to you.
Temporarily lend you the account.
Please check.

Do you know where to input id & pw in Mo PTT?

There are many mobile PTT BBS apps.
https://github.com/AdguardTeam/AdguardForAndroid/issues/1520#issue-264074297

Mo PTT & JPTT provides 2 connection methods.
1. Unsafe BBS Telnet connection
2. Safe SSH connection

I suggest that you install the 2 apps.
And you use **local VPN mode for the test.**
Please don't use local proxy mode.

JPTT:
https://play.google.com/store/apps/details?id=com.joshua.jptt

Mo PTT:
https://play.google.com/store/apps/details?id=mong.moptt&hl=zh_TW

If having a PTT BBS account...
With default port 443 and AdGuard VPN, JPTT needs to wait for around 2 minutes for login.
With default port 443 and AdGuard VPN, Mo PTT will finally show connection error after some time.

Some other mobile PTT BBS apps set SSH connection as the only way to PTT, such like PiTT app, PTT+ app and so on.
We don't use this kind of mobile PTT BBS apps for our test because AdGuard is unable to filter SSH connection.

We need to solve the issue between Unsafe BBS Telnet connection and AdGuard VPN.

Those PTT apps are Traditional Chinese interface.
Please feel free to contact with me if needing my help.

B/R

---

@lancelot-moon commented on [Fri Mar 02 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370004223)

@Revertron 

You can first install and run PCMan in Windows, which maybe helps you to be familiar with PTT BBS.

PCMan:
http://pcman.ptt.cc/downloads.html

Address: ptt.cc
Don't input guest since you already have an account.
(Guests are difficult to login into PTT BBS because the amount of guests is limited.)
Input the id & pw which I lend you for login into PTT.

login interface:
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/o9oAAVr.png)
</details><br/>

PTT BBS operation interface:
If having an account, you can use more complete features there.
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/OBtj45f.png)
</details><br/>

Such like VK is the biggest social community in Russia.
PTT BBS is the biggest social community in Taiwan.

Many Taiwanese net friends started to know AdGuard there because some other net friends and I introduced AdGuard on PTT BBS.

Most AdGuard's Taiwanese users have the same experience.
When using AdGuard VPN, they're unable to login into PTT BBS.
If they also don't know the information (change the port number or use SSH for login), they think AdGuard makes trouble.

That's why I hope that AdGuard can completely solve the issue between Unsafe BBS Telnet connection and AdGuard VPN.
**Please let users not need to change the port number or use SSH for login PTT BBS.**

Thanks.

---

@ameshkov commented on [Sun Mar 04 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370217415)

@Revertron as I understand, when the port is set to 443, AG tries to filter it as an HTTPS connection. The problem is that inside of the tunnel the protocol is not HTTP, and AG simply breaks it.

We need to figure out how to solve that. The easiest solution would be to add affected domains to the HTTPS exclusions list.

---

@lancelot-moon commented on [Sun Mar 04 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370220959)

**Adding ptt.cc in the HTTPS exclusions list is useless.
Disabling HTTPS filtering is also useless.**
The Telnet address is "telnet://ptt.cc", which is not HTTPS protocol.

@ameshkov 
I had lent a PTT BBS ID & Password to Revertron.
You can also use that account together.
**Multi-login is available.**
When multi-login, it will show the messages.
Please input "n" and then click "enter" on keyboard.
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/iG5T2XS.png)
</details><br/>

I suggest that you use PCMan or other BBS client apps for login into PTT BBS on your Windows PC.
PTT BBS address: telnet://ptt.cc
It's helpful to let you know what BBS is.

An experimental safe connection to PTT BBS, compared with SSH.
It's WebSocket Protocol.
We won't need a BBS app.
Only use a browser for login into PTT BBS.
https://term.ptt.cc

Some Taiwanese people are trying new safe connection to PTT BBS.
https://github.com/iamchucky/PttChrome

PttChrome:
https://chrome.google.com/webstore/detail/pttchrome/hhnlfapopmaimdlldbknjdgekpgffmbo

PTT Telnet BBS has the 3 associated servers.
telnet://ptt.cc <- (批踢踢實業坊) (Main server. Almost all of its users only visit this server.)
telnet://ptt2.cc <- (批踢踢兔) (different place, only a few users.)
telnet://ptt3.cc <- (批踢踢參) (different place, only a few users.)

The allowed ports for connecting to PTT BBS are only 23, 443 and 3000~3010.
For example:
telnet://ptt.cc:443
telnet://ptt.cc:3005

PTT BBS doesn't provide an official app.
PTT BBS itself has no ad.
Because there are several tens thousand (usually one hundred thousand) net friends on PTT BBS at the same time, some designers design client apps for login into PTT BBS.
They insert ads API in apps for earning money.

Most Taiwanese use Mo PTT or JPTT for login into PTT BBS on Android/iOS.
JPTT:
https://play.google.com/store/apps/details?id=com.joshua.jptt

Mo PTT:
https://play.google.com/store/apps/details?id=mong.moptt&hl=zh_TW

With AdGuard **local VPN mode**, I don't know why there are different results when using Mo PTT app for login into PTT BBS as a member or a guest.

On iOS, mobile PTT BBS apps have no connection issue with AdGuard Pro via port 443.
On Android, mobile PTT BBS apps have the connection issue with AdGuard via port 443 when using unsafe BBS Telnet protocol for connecting.
Constantly try to connect.
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/XqJvy7r.png)
</details><br/>

Finally, Mo PTT showed the error messages.
**"Oh! The server has no response, please try again later."**
<details><summary>Screenshot:</summary>

![image](https://i.imgur.com/ZgBHZEc.png)
</details><br/>

If you use JPTT, you can finally login into PTT BBS with AdGuard, **but you need to wait for 1.5~2 minutes for login**....

These all are abnormal situation.
**You need to design that AdGuard doesn't filter BBS Telnet protocol.**

---

@Revertron commented on [Mon Mar 05 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370473375)

I have a great news for you, @lancelot-moon! We've nailed this problem. It seems that this BBS software is waiting from the server the first request instead of sending it's request to the server first as do all the "normal" apps and browsers.
The fix will be in next Nightly, very soon :)

---

@lancelot-moon commented on [Mon Mar 05 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370487086)

Wow, so cool!!!
The issue tormented AdGuard's Taiwanese users for a long time.
**You're superman!** m(_  _)m
(*￣▽￣)/‧☆*"`'*-.,_,.-*'`"*-.,_☆ 

**Please test your method in not only Mo PTT but also JPTT since the 2 apps give the different results of trying to connect to PTT BBS with AdGuard local VPN.**
Mo PTT: Oh! The server has no response, please try again later.
JPTT:  Need to wait for 1.5~2 minutes for login.

I will try the next Nightly version.
Thank you very much!

---

@lancelot-moon commented on [Tue Mar 06 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370913517)

Hi,
I saw that AdGuard beta v2.11.31 is released.
Have you solve the BBS issue in AdGuard beta v2.11.31?
Or do I need to wait next AdGuard Nightly version?
Which is newer about AdGuard v2.11.32η and AdGuard beta v2.11.31?
Thanks.

---

@Revertron commented on [Tue Mar 06 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-370925275)

As I've said earlier, the fix will be in the next Nightly, sorry :-\

---

@lancelot-moon commented on [Wed Mar 07 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-371215981)

Sir,
Thank you very much!
I tried v2.11.43η .
Now, mobile PTT BBS apps can normally login via port 443 with AdGuard local VPN.

But their speed of login are delayed.
Without AdGuard local VPN:
JPTT takes 1 second for login.
MoPTT takes 2 seconds for login.

With AdGuard local VPN:
They need about 6 seconds for login.

Is it possible to improve the time of login into PTT BBS with AdGuard local VPN?

Thanks a lot!

---

@Revertron commented on [Wed Mar 07 2018](https://github.com/AdguardTeam/AdguardForAndroid/issues/1601#issuecomment-371218123)

We are aware of that delay. Unfortunately, this delay is for all the apps and we can not do any exception. At least for now. If we make this delay very short, some ads can slip through.

**Comment by lancelot-moon** [comment]
@Revertron
> We are aware of that delay. Unfortunately, this delay is for all the apps and we can not do any exception. At least for now. If we make this delay very short, some ads can slip through.

Honestly, I don't fully understand Revertron's reply about **the delay time** since I supposed that AdGuard was already ready for filtering all of the internet traffic after enabling AdGuard Protection.
I supposed AdGuard had done its preparation.
Except for filtering websites, when I used other apps with AdGuard local VPN, I didn't feel so obvious delay time.

And then PTT BBS itself doesn't have ads.
The delay time is unnecessary for filtering BBS since BBS has no ad.
Only those mobile PTT BBS apps' designers make ads API in the apps.
Those ads are from external ad servers, not from PTT BBS itself.
https://github.com/AdguardTeam/AdguardFilters/issues/9972#issuecomment-371544106

For example, Mo PTT app's designers built the private server which provides some additional features for  Mo PTT app, such like hot articles, the view of voting in all of boards and users' avatar photo...etc.
PTT BBS itself doesn't have those additional features.

The domain of Mo PTT app's private server is azurewebsites.net.
The URL, moptt.azurewebsites.net/api/pttboard/update, calls external ad servers.
Most ads from external ad servers can be blocked by AdGuard.
However, about some ads which are called by the URL, moptt.azurewebsites.net/api/pttboard/update, it seems that they will **slip through** AdGuard's HTTPS filtering if we don't block moptt.azurewebsites.net.

For AdGuard for Android with root access, we can just block the URL.
For AdGuard for Andoid with non-root access or AdGuard Pro for iOS, we can just block the domain, moptt.azurewebsites.net.

If blocking moptt.azurewebsites.net in the blacklist, it doesn't affect the login into PTT BBS and any its own features of PTT BBS, but those additional features by Mo PTT app will become unavailable.

B/R

**Comment by Revertron** [comment]
@lancelot-moon I'll try to put it simple:
There are two types of connections. First type is simple, like in browsers and most of apps - client makes a connection, then sends a request.
The second type (that uses Mo PTT for example) is somewhat more complicated - client makes a connection and waits for the first data from the server.
The second type is like fall-back type, like an exception. Because of it, when we detect that client is not sending data, we initiate simple tunneling. But we _wait_ for the data some time, hence the delay.
And the simple tunneling mode is not filtering ads. And the delay, when we wait for the first data from the client, is used for _all_ apps. If we lower the delay, some apps may (because of some lags for example) start to fall-back to simple tunneling mode and will not be filtered.
Bottom line is, that we wont do some individual delay lowering for Mo PTT or something else.

**Comment by TPS** [comment]
Since others are, can this delay be made configurable in Advanced Settings, @Revertron?

**Comment by sfionov** [comment]
CoreLibs doesn't need any pause to determine protocol - when first packet is from client and it is ClientHello - it starts HTTPS filtering. Else, if first packet is from server, or first packet from client is not ClientHello, it bypasses that connection.

**Comment by ameshkov** [comment]
@sfionov yeah, but we'd better re-test everything when the corelibs-integrated version passes the code review.

**Comment by ameshkov** [comment]
Confirmed that it works okay in the CoreLibs integrated Android build
